# Week 9.2 Assignment
- Joshua Burden
- Bellevue University
- DSC550 Data Mining
- Dr. Brett Werner
- 10/30/2022

# Import the dataset and ensure that it loaded properly.

In [52]:
#import necessary packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split #used to split data into training/test sets
from sklearn.metrics import plot_confusion_matrix
import sklearn.metrics as metrics
from sklearn.metrics import confusion_matrix
from yellowbrick.classifier import ConfusionMatrix
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

In [77]:

# Import the pandas library
import pandas as pd

# Load the datafile
df = pd.read_csv(r'./DATA/Loan_Train.csv')

# Validate the data loaded correctly
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y


# Prepare the data for modeling by performing the following steps:


# Drop the column “Load_ID.”


In [78]:
# Drop the "Load_ID" column
df = df.drop(['Loan_ID'], axis=1)

# Validate the transformation was successful
df.head()

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0   Male      No          0      Graduate            No             5849   
1   Male     Yes          1      Graduate            No             4583   
2   Male     Yes          0      Graduate           Yes             3000   
3   Male     Yes          0  Not Graduate            No             2583   
4   Male      No          0      Graduate            No             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0         NaN             360.0             1.0   
1             1508.0       128.0             360.0             1.0   
2                0.0        66.0             360.0             1.0   
3             2358.0       120.0             360.0             1.0   
4                0.0       141.0             360.0             1.0   

  Property_Area Loan_Status  
0         Urban           Y  
1         Rural           N  
2         Urban           Y  
3         Urban           Y  
4         Urban           Y

# Drop any rows with missing data.


In [55]:
# View the shape of the df
df.shape
(614, 12)


(614, 12)

In [79]:
# Drop any rows with any missing data
df = df.dropna(how='any')


In [80]:
# View the new shape of the df
df.shape

(480, 12)

In [81]:
# Obtain the NA counts by column
nacounts = df.isna().sum()

# View the NA counts by column
nacounts

Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

# Convert the categorical features into dummy variables.


In [82]:
# Find numerical columns
numerical_columns = df.select_dtypes(include=['number']).columns

# Find categorical columns
cols = df.columns
categorical_columns = list(set(cols)-set(numerical_columns))

In [83]:
# Use the get_dummies function to create the dummy variables
df_dummies = pd.get_dummies(df, columns=categorical_columns)

# View the first 5 rows of the dataframe
df_dummies.head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History  Gender_Female  Gender_Male  Married_No  Married_Yes  \
1             1.0              0            1           0            1   
2             1.0              0            1           0            1   
3             1.0              0            1           0            1   
4             1.0              0            1           1            0   
5             1.0              0            1           0            1   

   Education_Graduate  ...  Property_Area_Semiurban  Property_Area_Urban  \
1                   1  ...                        0                    0   
2                   1  ...                        0                    1   
3                   0  ...                        0                    1   
4                   1  ...                        0                    1   
5                   1  ...                        0                    1   

   Loan_Status_N  Loan_Status_Y  Dependents_0  Dependents_1  Dependents_2  \
1              1              0             0             1             0   
2              0              1             1             0             0   
3              0              1             1             0             0   
4              0              1             1             0             0   
5              0              1             0             0             1   

   Dependents_3+  Self_Employed_No  Self_Employed_Yes  
1              0                 1                  0  
2              0                 0                  1  
3              0                 1                  0  
4              0                 1                  0  
5              0                 0                  1  

[5 rows x 22 columns]

In [84]:
# Drop columns that are redundant information (dummy variables created on binary fields)
df_dummies = df_dummies.drop(['Education_Not Graduate'], axis=1)
df_dummies = df_dummies.drop(['Gender_Female'], axis=1)
df_dummies = df_dummies.drop(['Self_Employed_No'], axis=1)
df_dummies = df_dummies.drop(['Married_No'], axis=1)
df_dummies = df_dummies.drop(['Loan_Status_N'], axis=1)

In [85]:
df_dummies.head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History  Gender_Male  Married_Yes  Education_Graduate  \
1             1.0            1            1                   1   
2             1.0            1            1                   1   
3             1.0            1            1                   0   
4             1.0            1            0                   1   
5             1.0            1            1                   1   

   Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
1                    1                        0                    0   
2                    0                        0                    1   
3                    0                        0                    1   
4                    0                        0                    1   
5                    0                        0                    1   

   Loan_Status_Y  Dependents_0  Dependents_1  Dependents_2  Dependents_3+  \
1              0             0             1             0              0   
2              1             1             0             0              0   
3              1             1             0             0              0   
4              1             1             0             0              0   
5              1             0             0             1              0   

   Self_Employed_Yes  
1                  0  
2                  1  
3                  0  
4                  0  
5                  1

# Split the data into a training and test set, where the “Loan_Status” column is the target.


In [86]:
# Create x & y arrays
x = df_dummies.drop('Loan_Status_Y', axis=1)
y = df_dummies['Loan_Status_Y']

# Import library
from sklearn.model_selection import train_test_split

# Create training & test datasets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2)



# Create a pipeline with a min-max scaler and a KNN classifier (see section 15.3 in the Machine Learning with Python Cookbook).


In [87]:
# Load libraries
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.model_selection import GridSearchCV

# Create standardizer
standardizer = StandardScaler()

# Standardize features
x_train_s = standardizer.fit_transform(x_train)
x_test_s = standardizer.transform(x_test)

# Create a KNN classifier
knn = KNeighborsClassifier(n_neighbors=4, n_jobs=-1)

# Create a pipeline
pipe = Pipeline([("standardizer", standardizer), ("knn", knn)])

# Fit a default KNN classifier to the data with this pipeline. Report the model accuracy on the test set. Note: Fitting a pipeline model works just like fitting a regular model.


In [88]:
# Fit the pipeline
pipe.fit(x_train_s, y_train)

Pipeline(steps=[('standardizer', StandardScaler()),
                ('knn', KNeighborsClassifier(n_jobs=-1, n_neighbors=4))])

In [89]:
# Load libraries
from sklearn.metrics import accuracy_score

# Build predictions
pipe_pred = pipe.predict(x_test_s)

# Calculate accuracy
accuracy_score(y_test, pipe_pred)

0.71875

# Create a search space for your KNN classifier where your “n_neighbors” parameter varies from 1 to 10. (see section 15.3 in the Machine Learning with Python Cookbook).


In [90]:
search_space = [{"knn__n_neighbors": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}]

# Fit a grid search with your pipeline, search space, and 5-fold cross-validation to find the best value for the “n_neighbors” parameter.


In [91]:
# Create grid search
classifier = GridSearchCV(pipe, search_space, cv=5, verbose=0).fit(x_train_s, y_train)


In [92]:
# Find the best value
classifier.best_estimator_.get_params()["knn__n_neighbors"]

10

# Find the accuracy of the grid search best model on the test set. Note: It is possible that this will not be an improvement over the default model, but likely it will be.


In [93]:
# Build predictions
classifier_pred = classifier.predict(x_test_s)

# Calculate accuracy
accuracy_score(y_test, classifier_pred)

0.7291666666666666

# Now, repeat steps 6 and 7 with the same pipeline, but expand your search space to include logistic regression and random forest models with the hyperparameter values in section 12.3 of the Machine Learning with Python Cookbook.


In [94]:
# Load libraries
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

# Build new search space
search_space2 = [{"classifier": [LogisticRegression()], 
                  "classifier__penalty": ['l2'],
                  "classifier__C": np.logspace(0, 4, 10)}, 
                 {"classifier": [RandomForestClassifier()], 
                  "classifier__n_estimators": [10, 100, 1000], 
                  "classifier__max_features": [1, 2, 3]}]

# Build new pipeline
pipe2 = Pipeline([("classifier", RandomForestClassifier())])

# Fit the pipeline
pipe2.fit(x_train_s, y_train)

Pipeline(steps=[('classifier', RandomForestClassifier())])

In [95]:
# grid search 
classifier2 = GridSearchCV(pipe2, search_space2, cv=5, verbose=0).fit(x_train_s, y_train)

# What are the best model and hyperparameters found in the grid search? Find the accuracy of this model on the test set.


In [96]:
classifier2.best_estimator_.get_params()["classifier"]

RandomForestClassifier(max_features=3, n_estimators=1000)

In [97]:
# Build predictions
classifier2_pred = classifier2.predict(x_test_s)

# Calculate accuracy
accuracy_score(y_test, classifier2_pred)

0.8229166666666666

# Summarize your results.

Using hyperparameter tuning through the KNN model our accuracy increased from 71% to 73%. When increasing the grid search into different models we are able to get an 7% increase to 82%. 